### PR TITLE
Implements Ethereum connected sites preference on Android

### PR DESCRIPTION
**Pull request by SergeyZhukovsky**
Resolves https://github.com/brave/brave-browser/issues/22805

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Screenshot_20220527-092448_Brave - Debug

### CODE REVIEW

File: components/permissions/contexts/brave_wallet_permission_context.cc [hunk 1]
```diff
@@ -334,4 +334,44 @@ bool BraveWalletPermissionContext::ResetPermission(
...
+// static
+bool BraveWalletPermissionContext::ResetWebSitePermission(
+    ContentSettingsType content_settings_type,
+    content::BrowserContext* context,
+    const std::string& formed_website) {
+  auto* permission_manager = static_cast<permissions::BravePermissionManager*>(
+      permissions::PermissionsClient::Get()->GetPermissionManager(context));
+  if (!permission_manager)
+    return false;
+  GURL url(formed_website);
```
Comment (yrliou) [inline components/permissions/contexts/brave_wallet_permission_context.cc:370]: I think we should check `url.is_valid()` and return false when it's not valid. This should be covered in unit test too.

File: android/java/org/chromium/chrome/browser/crypto_wallet/modal/BraveWalletPanel.java [hunk 2]
```diff
@@ -127,7 +127,11 @@ public void showPopupMenu(){
                 mBraveWalletPanelServices.getKeyringService().lock();
                 dismiss();
             } else if (item.getItemId() == R.id.action_connected_sites) {
-                //TODO needs to be implemented in another ticket
+                BraveActivity activity = BraveActivity.getBraveActivity();
+                if (activity != null) {
+                    activity.openBraveConnectedSitesSettings();
+                    dismiss();
+                }
                 dismiss();
```
Comment (Pavneet-Sing) [inline android/java/org/chromium/chrome/browser/crypto_wallet/modal/BraveWalletPanel.java:135]: nit: Can remove either of the `dismiss` call.

File: android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesListAdapter.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        mWebSites = webSites;
+        mDelegate = delegate;
+    }
+
+    public void setWebSites(String[] webSites) {
+        mWebSites = webSites;
+    }
+
+    @Override
+    public @NonNull BraveWalletEthereumConnectedSitesListAdapter.ViewHolder onCreateViewHolder(
```
Comment (Pavneet-Sing) [inline android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesListAdapter.java:41]: nit: can remove `BraveWalletEthereumConnectedSitesListAdapter.`

File: android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesListAdapter.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        LayoutInflater inflater = LayoutInflater.from(mContext);
+        View view =
+                inflater.inflate(R.layout.brave_wallet_connected_sites_list_item, parent, false);
+
+        return new ViewHolder(view);
+    }
+
+    @Override
+    public void onBindViewHolder(
+            @NonNull BraveWalletEthereumConnectedSitesListAdapter.ViewHolder holder, int position) {
```
Comment (Pavneet-Sing) [inline android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesListAdapter.java:53]: nit: can remove `BraveWalletEthereumConnectedSitesListAdapter.`

File: android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesListAdapter.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public static class ViewHolder extends RecyclerView.ViewHolder {
+        public TextView webSite;
+        public ImageView iconTrash;
+
+        public ViewHolder(View itemView) {
+            super(itemView);
+            this.webSite = itemView.findViewById(R.id.web_site);
+            this.iconTrash = itemView.findViewById(R.id.icon_trash);
```
Comment (Pavneet-Sing) [inline android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesListAdapter.java:74]: nit: can remove `this.`

File: android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesPreference.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        super(context, attrs);
+    }
+
+    @Override
+    public void onBindViewHolder(PreferenceViewHolder holder) {
+        super.onBindViewHolder(holder);
+
+        initBraveWalletService();
+
+        mRecyclerView = (RecyclerView) holder.findViewById(R.id.connected_sites_list);
```
Comment (Pavneet-Sing) [inline android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesPreference.java:43]: nit: do we need the casting here `(RecyclerView)`? maybe can remove

File: android/java/res/xml/brave_ethereum_preferences.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) 2020 The Brave Authors. All rights reserved.
```
Comment (Pavneet-Sing) [inline android/java/res/xml/brave_ethereum_preferences.xml:2]: 2022

File: android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesPreference.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        LinearLayoutManager layoutMangager = new LinearLayoutManager(getContext());
+        mRecyclerView.setLayoutManager(layoutMangager);
```
Comment (Pavneet-Sing) [inline android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesPreference.java:45]: nit: can move to XML with `app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"`

**Review by Pavneet-Sing — APPROVED** [review]
++ for android/java

File: android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesPreference.java (hunk 6 above)
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/site_settings/BraveWalletEthereumConnectedSitesPreference.java:43]: `error: incompatible types: View cannot be converted to RecyclerView`

File: components/permissions/contexts/brave_wallet_permission_context.cc [hunk 9]
```diff
@@ -334,4 +334,44 @@ bool BraveWalletPermissionContext::ResetPermission(
...
+  map->GetSettingsForOneType(content_settings_type, &settings);
+
+  std::vector<std::string> result;
+  for (const auto& setting : settings) {
+    if (setting.primary_pattern.GetScheme() !=
+            ContentSettingsPattern::SchemeType::SCHEME_HTTP &&
+        setting.primary_pattern.GetScheme() !=
+            ContentSettingsPattern::SchemeType::SCHEME_HTTPS) {
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply components/permissions/contexts/brave_wallet_permission_context.cc:354]: That way I excluded a wildcard pattern `*`. But perhaps it's better to just check on `CONTENT_SETTING_ALLOW`. I'm remaking it to return everything that has `CONTENT_SETTING_ALLOW`

File: android/java/org/chromium/chrome/browser/app/BraveActivity.java [hunk 10]
```diff
@@ -991,7 +991,7 @@ public void openBraveWallet(boolean fromDapp) {
     }
 
     public void viewOnBlockExplorer(String address) {
-        Utils.openAddress("/address/"+address, mJsonRpcService, this);
+        Utils.openAddress("/address/" + address, mJsonRpcService, this);
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/app/BraveActivity.java:994]: not related to the PR lint problems

File: android/java/org/chromium/chrome/browser/crypto_wallet/modal/BraveWalletPanel.java [hunk 11]
```diff
@@ -119,7 +119,7 @@ public void onDismiss() {
         setUpViews();
     }
 
-    public void showPopupMenu(){
+    public void showPopupMenu() {
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/modal/BraveWalletPanel.java:122]: not related to the PR lint problems

File: android/java/org/chromium/chrome/browser/crypto_wallet/modal/BraveWalletPanel.java [hunk 12]
```diff
@@ -141,10 +141,11 @@ public void showPopupMenu(){
...
                 if (activity != null) {
-                    mBraveWalletPanelServices.getKeyringService().getSelectedAccount(CoinType.ETH, address -> {
-                        activity.viewOnBlockExplorer(address);
-                        dismiss();
-                    });
+                    mBraveWalletPanelServices.getKeyringService().getSelectedAccount(
+                            CoinType.ETH, address -> {
+                                activity.viewOnBlockExplorer(address);
+                                dismiss();
+                            });
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/modal/BraveWalletPanel.java:148]: not related to the PR lint problems

File: components/permissions/contexts/brave_wallet_permission_context.cc [hunk 13]
```diff
@@ -334,4 +334,44 @@ bool BraveWalletPermissionContext::ResetPermission(
   return true;
 }
 
+// static
+std::vector<std::string>
+BraveWalletPermissionContext::GetWebSitesWithPermission(
```
[the remark this answers is not in the document]
Comment (SergeyZhukovsky) [reply components/permissions/contexts/brave_wallet_permission_context.cc:339]: done

File: components/permissions/contexts/brave_wallet_permission_context.cc (hunk 1 above)
Comment (SergeyZhukovsky) [reply components/permissions/contexts/brave_wallet_permission_context.cc:365]: done

**Review by yrliou — APPROVED** [review]
++ for c++ and mojom

File: components/permissions/contexts/brave_wallet_permission_context.cc [hunk 14]
```diff
@@ -334,4 +334,41 @@ bool BraveWalletPermissionContext::ResetPermission(
...
+      PermissionsClient::Get()->GetSettingsMap(context);
+  ContentSettingsForOneType settings;
+  map->GetSettingsForOneType(content_settings_type, &settings);
+
+  std::vector<std::string> result;
+  for (const auto& setting : settings) {
+    if (setting.GetContentSetting() != CONTENT_SETTING_ALLOW) {
+      continue;
+    }
+    result.push_back(setting.primary_pattern.ToString());
```
Comment (yrliou) [inline components/permissions/contexts/brave_wallet_permission_context.cc:352]: nit: can be simplified as
```
if (setting.GetContentSetting() == CONTENT_SETTING_ALLOW) {
  result.push_back(setting.primary_pattern.ToString());
}
```

Comment (SergeyZhukovsky) [reply components/permissions/contexts/brave_wallet_permission_context.cc:352]: yeah, I also noticed that, after CI has already started. Not a big deal imho, I just don't want to run it one more time.

Comment (yrliou) [reply components/permissions/contexts/brave_wallet_permission_context.cc:352]: Okay, I'm fine with merging as is.

**Pull request merged**